cabal-install/src/Distribution/Client/InLibrary.hs

```diff
@@ -0,0 +1,352 @@
+  :: ProgramDb
+  -> ElaboratedSharedConfig
+  -> ElaboratedReadyPackage
+  -> [String]
+  -- ^ targets
+  -> LibraryConfigureInputs
+libraryConfigureInputsFromElabPackage
+  progDb
+  -- NB: don't use the ProgramDb from the ElaboratedSharedConfig;
+  -- that one is only for the compiler itself and not for the package.
+  ElaboratedSharedConfig
+    { pkgConfigPlatform = platform
+    , pkgConfigCompiler = compiler
+    }
+  (ReadyPackage pkg)
+  userTargets =
+    LibraryConfigureInputs
+      { compiler
+      , platform
+      , buildType = PD.buildType pkgDescr
+      , compRequested =
+          case elabPkgOrComp pkg of
+            ElabComponent elabComp
+              | Just elabCompNm <- compComponentName elabComp ->
+                  Just elabCompNm
+            _ -> Nothing
+      , localBuildConfig =
+          LBC.LocalBuildConfig
+            { LBC.extraConfigArgs = userTargets
+            , LBC.withPrograms = progDb
+            , LBC.withBuildOptions = elabBuildOptions pkg
+            }
+      , packageDBStack = elabBuildPackageDBStack pkg
+      , packageDescription = pkgDescr
+      , gPackageDescription = gpkgDescr
+      , flagAssignment = elabFlagAssignment pkg
+      }
+    where
+      pkgDescr = elabPkgDescription pkg
+      gpkgDescr = elabGPkgDescription pkg
+
+configure
+  :: LibraryConfigureInputs
+  -> Cabal.ConfigFlags
+  -> IO LocalBuildInfo
+configure
+  LibraryConfigureInputs
+    { platform
+    , compiler
+    , buildType = bt
+    , compRequested = mbComp
+    , localBuildConfig = lbc0
+    , packageDBStack = packageDBs
+    , packageDescription = pkgDesc
+    , gPackageDescription = gpkgDescr
+    , flagAssignment
+    }
+  cfg =
+    -- SetupHooks TODO: the following code should not live in cabal-install.
+    -- We should be able to directly call into the library,
+    -- similar to what we do for other phases (see e.g. inLibraryBuild).
+    --
+    -- The issue is mainly about 'finalizeAndConfigurePackage' vs 'configurePackage'.
+    do
+      let verbosity = Cabal.fromFlag $ Cabal.configVerbosity cfg
+          mbWorkDir = Cabal.flagToMaybe $ Cabal.configWorkingDir cfg
+          distPref = Cabal.fromFlag $ Cabal.configDistPref cfg
+          confHooks = configureHooks $ ExternalHooksExe.buildTypeSetupHooks mbWorkDir distPref bt
+
+      -- Configure package
+      let pkgId :: PD.PackageIdentifier
+          pkgId = PD.package pkgDesc
+      case mbComp of
+        Nothing -> setupMessage verbosity "Configuring" pkgId
+        Just cname ->
+          setupMessage'
+            verbosity
+            "Configuring"
+            pkgId
+            cname
+            (Just (Cabal.configInstantiateWith cfg))
+
+      -- SetupHooks TODO: we should avoid re-doing package-wide things
+      -- over and over in the per-component world, e.g.
+      --   cabal build comp1 && cabal build comp2
+      -- should only run the per-package configuration (including hooks) a single time.
```

Comment:
What about this?

Reply:
I think this would require an architectural change to `buildAndRegisterUnpackedPackage` to allow it to handle individual components and recompilation checking of individual components.

Reply:
I see. Given this is not in the critical path for this PR I think it is fine to leave it.